torch_xla/csrc/xla_sharding_util.cpp

```diff
@@ -311,32 +311,49 @@ std::vector<at::Tensor> ShardingUtil::ShardTensor(
         int64_t v_partition = tensor.sizes()[4] / tile_shape[4] +
                               (tensor.sizes()[4] % tile_shape[4] != 0);
         shard = tensor.index(
-            {at::indexing::Slice((core / tile_shape[1] / tile_shape[2] /
+            {at::indexing::Slice((i / tile_shape[1] / tile_shape[2] /
                                   tile_shape[3] / tile_shape[4]) *
                                      x_partition,
-                                 (core / tile_shape[1] / tile_shape[2] /
+                                 (i / tile_shape[1] / tile_shape[2] /
                                       tile_shape[3] / tile_shape[4] +
                                   1) *
                                      x_partition),
              at::indexing::Slice(
-                 (core / tile_shape[2] / tile_shape[3] / tile_shape[4]) *
+                 (i / tile_shape[2] / tile_shape[3] / tile_shape[4]) *
                      y_partition,
-                 (core / tile_shape[2] / tile_shape[3] / tile_shape[4] + 1) *
+                 (i / tile_shape[2] / tile_shape[3] / tile_shape[4] + 1) *
                      y_partition),
              at::indexing::Slice(
-                 (core / tile_shape[3] / tile_shape[4]) * z_partition,
-                 (core / tile_shape[3] / tile_shape[4] + 1) * z_partition),
-             at::indexing::Slice((core / tile_shape[4]) * w_partition,
-                                 (core / tile_shape[4] + 1) * w_partition),
-             at::indexing::Slice((core % tile_shape[4]) * v_partition,
-                                 (core % tile_shape[4] + 1) * v_partition)});
+                 (i / tile_shape[3] / tile_shape[4]) * z_partition,
+                 (i / tile_shape[3] / tile_shape[4] + 1) * z_partition),
+             at::indexing::Slice((i / tile_shape[4]) * w_partition,
+                                 (i / tile_shape[4] + 1) * w_partition),
+             at::indexing::Slice((i % tile_shape[4]) * v_partition,
+                                 (i % tile_shape[4] + 1) * v_partition)});
       }
+      int64_t core = sharding.tile_assignment_devices()[i];
       shards[core] = shard.contiguous(at::MemoryFormat::Contiguous);
     }
   } else if ((sharding.type() == xla::OpSharding::MANUAL) ||
              (sharding.type() == xla::OpSharding::TUPLE)) {
     TF_LOG(ERROR) << "Unsupported OpSharidng type " << sharding.type();
   }
+
+  // Zero-pad to the right to ensure the sizes are even
+  if (shards.size() > 0 && padded) {
+    for (size_t i = 1; i < shards.size(); ++i) {
+      std::vector<long> pads;
+      for (size_t j = 0; j < shards[i].sizes().size(); ++j) {
+        XLA_CHECK_GE(shards[0].sizes().at(j), shards[i].sizes().at(j));
+        pads.push_back(shards[0].sizes().at(j) - shards[i].sizes().at(j));
+        pads.push_back(0);  // no padding on lhs
+      }
+      // Padding starts from the last dimension
+      std::reverse(pads.begin(), pads.end());
+      shards[i] = at::constant_pad_nd(
```

Comment:
Just for my understanding - if we apply padding to the shards, does the padding need to be removed when reassembling the tensor?

Reply:
Good question, we don't need to reassemble for the computation execution, but we will need to account for the padding when we do (e.g., provide access to the shards, etc.). Let's create a task to track.